### PR TITLE
NP-45137-fetch-nvi-candidate

### PR DESCRIPTION
**Pull request by hackon**

### CODE REVIEW

File: docs/openapi.yaml [hunk 1]
```diff
@@ -39,6 +39,48 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/Problem"
+  /candidates/{publicationIdentifier}:
```
Comment (truhacevkir) [inline docs/openapi.yaml:42]: maybe candidateIdentifier? since publicationIdentifier = candidateIdentifier

File: docs/openapi.yaml [hunk 2]
```diff
@@ -39,6 +39,48 @@ paths:
...
+    get:
+      description: 'Get scientific index candidate'
+      parameters:
+        - in: path
+          name: publicationIdentifier
```
Comment (truhacevkir) [inline docs/openapi.yaml:47]: same here

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandler.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        var publicationIdentifier =
+            attempt(() -> requestInfo.getPathParameter("publicationIdentifier"))
+                .map(URI::create)
+                .orElseThrow((failure) ->
+                                 new BadRequestException("Missing publicationIdentifier"));
+        return service.getCandidate(publicationIdentifier)
+                   .map(FetchCandidateResponse::fromCandidate)
+                   .orElseThrow(() -> new NotFoundException(
+                       "No candidate found for " + publicationIdentifier.toString()));
+    }
```
Comment (truhacevkir) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandler.java:36]: Can be shortened to:
```
return attempt(() -> requestInfo.getPathParameter(CANDIDATE_IDENTIFIER))
                   .map(URI::create)
                   .map(service::getCandidate)
                   .map(FetchCandidateResponse::fromCandidate)
                   .orElseThrow(() -> notFoundException(requestInfo.getPathParameter(CANDIDATE_IDENTIFIER)));
```

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        return new FetchCandidateResponse(
+            candidate.publicationId(),
+            candidate.approvalStatuses()
+                .stream()
+                .map(as -> new ApprovalStatus(as.institutionId(), as.status(), as.finalizedBy(), as.finalizedDate()))
+                .toList(),
+            candidate.notes()
+                .stream()
+                .map(n -> new Note(n.user(), n.text(), n.createdDate()))
+                .toList()
```
Comment (truhacevkir) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java:27]: extract to methods

File: template.yaml [hunk 5]
```diff
@@ -274,4 +274,27 @@ Resources:
...
+      AutoPublishAlias: live
+      Environment:
+        Variables:
+          ALLOWED_ORIGIN: '*'
+          API_HOST: !Ref ApiDomain
+      Events:
+        GetEvent:
+          Type: Api
+          Properties:
+            Path: /candidates/{publicationIdentifier}
```
Comment (truhacevkir) [inline template.yaml:298]: candidateIdentifier?

File: nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    @Test
+    void shouldReturn404whenNoCandidateFound() throws IOException {
+
+        var publicationIdentifier = randomUri();
+        InputStream input = getInput(publicationIdentifier);
+        when(service.getCandidate(URI.create(""))).thenReturn(Optional.empty());
+        handler.handleRequest(input, output, CONTEXT);
+        GatewayResponse<FetchCandidateResponse> gatewayResponse = getGatewayResponse();
```
Comment (truhacevkir) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java:74]: var

File: nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        assertEquals(HttpStatus.SC_NOT_FOUND, gatewayResponse.getStatusCode());
+    }
+
+    @Test
+    void shouldReturnValidCandidateIfExists() throws IOException {
+
+        var publicationId = randomUri();
+        var candidate = getCandidate(publicationId);
+        when(service.getCandidate(publicationId)).thenReturn(Optional.of(candidate));
+        InputStream input = getInput(publicationId);
```
Comment (truhacevkir) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java:85]: var

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java [hunk 8]
```diff
@@ -29,6 +31,12 @@ public void upsertCandidate(CandidateEvaluatedMessage evaluatedCandidate) {
         }
     }
 
+    //TODO: Remove JacocoGenerated when other if/else cases are implemented
+    @JacocoGenerated
+    public Optional<Candidate> getCandidate(URI publicationIdentifier) {
```
Comment (truhacevkir) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/NviService.java:36]: Do we want to return Optional here? Maybe return Candidate and throw not found if does not exist?

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandler.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    public FetchNviCandidateHandler(NviService service, Environment environment) {
+        super(Void.class, environment);
+        this.service = service;
+    }
+
+    @Override
+    protected FetchCandidateResponse processInput(Void input, RequestInfo requestInfo, Context context)
+        throws ApiGatewayException {
+        return attempt(() -> requestInfo.getPathParameter(PARAM_CANDIDATE_IDENTIFIER))
+                   .map(URI::create)
```
Comment (monaullah) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandler.java:29]: Is path param candidateIdentifier a URI or UUID? In the tests you are testing with a URI, but in the open api spec it seems like it is an UUID?

Comment (hackon) [reply nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandler.java:29]: @brinxmat A penny for your thoughts.
.../nvi/candidate/d5fa52eb-96da-4bc1-add9-979cc0978dfa
.../nvi/candidate/http%3A%2F%2Fnva.sikt.no%2Fpublication%2Fd5fa52eb-96da-4bc1-add9-979cc0978dfa

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.List;
+import no.sikt.nva.nvi.common.model.business.Candidate;
+
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+@JsonSerialize
+public record FetchCandidateResponse(
+    URI publicationId,
+    List<ApprovalStatus> approvalStatuses,
+    List<Note> notes
+) {
```
Comment (brinxmat) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java:15]: Is this weird formatting, or are records just like this?

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            candidate.publicationId(),
+            mapToApprovalStatus(candidate),
+            mapToNotes(candidate)
+        );
+    }
+
+    private static List<Note> mapToNotes(Candidate candidate) {
+        return candidate.notes()
+                   .stream()
+                   .map(n -> new Note(n.user(), n.text(), n.createdDate()))
```
Comment (brinxmat) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java:28]: What's an _n_?

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+                   .toList();
+    }
+
+    private static List<ApprovalStatus> mapToApprovalStatus(Candidate candidate) {
+        return candidate.approvalStatuses()
+                   .stream()
+                   .map(as -> new ApprovalStatus(as.institutionId(), as.status(), as.finalizedBy(), as.finalizedDate()))
```
Comment (brinxmat) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java:35]: Ditto :)

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandler.java (hunk 9 above)
Comment (brinxmat) [reply nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandler.java:29]: They cost more than a penny :(

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/Note.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.time.Instant;
+import no.sikt.nva.nvi.common.model.business.Username;
+
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+@JsonSerialize
+public record Note(Username user,
+                   String text,
+                   Instant createdDate) {
```
Comment (brinxmat) [inline nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/Note.java:12]: Here I see the formatting I expected

File: nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    private static final Context CONTEXT = mock(Context.class);
+    private ByteArrayOutputStream output;
+    private FetchNviCandidateHandler handler;
+    private Environment environment;
+    private NviService service;
+
+    @BeforeEach
+    public void setUp() {
+        environment = mock(Environment.class);
+        when(environment.readEnv(ALLOWED_ORIGIN_ENV)).thenReturn("*");
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java:49]: Gradle env?

File: nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+        output = new ByteArrayOutputStream();
+        service = mock(NviService.class);
+        handler = new FetchNviCandidateHandler(service, environment);
+    }
+
+    @Test
+    void shouldReturn400whenMissingCandidateId() throws IOException {
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java:57]: The path parameter for candidate identifier?

File: nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+                        .withHeaders(Map.of(HttpHeader.ACCEPT.asString(), ContentType.APPLICATION_JSON.getMimeType()))
+                        .build();
+        handler.handleRequest(input, output, CONTEXT);
+        var gatewayResponse = getGatewayResponse();
+
+        assertEquals(HttpStatus.SC_NOT_FOUND, gatewayResponse.getStatusCode());
+    }
+
+    @Test
+    void shouldReturn404whenNoCandidateFound() throws IOException {
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java:69]: CandidateDoesNotExist

File: nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+        var candidateIdentifier = randomUri();
+        InputStream input = getInput(candidateIdentifier);
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java:72]: var

File: nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+        when(service.getCandidate(URI.create(""))).thenReturn(Optional.empty());
+        handler.handleRequest(input, output, CONTEXT);
+        var gatewayResponse = getGatewayResponse();
+
+        assertEquals(HttpStatus.SC_NOT_FOUND, gatewayResponse.getStatusCode());
+    }
+
+    @Test
+    void shouldReturnValidCandidateIfExists() throws IOException {
```
Comment (brinxmat) [inline nvi-rest/src/test/java/no/sikt/nva/nvi/fetch/FetchNviCandidateHandlerTest.java:81]: Two things here:

- shouldReturnCandidateWhenExistingCandidateIsRequested
- valid…you check one thing…should we be validating candidates on output or input? We probably need to verify the structure of candidates in some test…we have done this previously with does-not-contain-empty-or-null-values-or-whatever-it-is-called

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java (hunk 10 above)
Comment (hackon) [reply nvi-rest/src/main/java/no/sikt/nva/nvi/fetch/FetchCandidateResponse.java:15]: Preferred formating?
